lib/debug-log.js

```diff
@@ -0,0 +1,325 @@
+    if (_.isNil(opts.fd)) {
+      const lockfilePath = path.join(tempDirpath, `${basename}.lock`);
+      const originalUnlock = await lock(logPath, {...BASE_LOCK_OPTS, lockfilePath});
+      const wrappedUnlock = async () => {
+        try {
+          await originalUnlock();
+          log.debug(`Unlocked ${logPath}`);
+        } catch (err) {
+          // todo handle better
+          log.warn(err);
+        }
+      };
+      return wrappedUnlock;
+    }
+    return () => B.resolve();
+  }
+
+  /**
+   * Unlocks lockfile and tries not to leak memory
+   */
+  async #cleanup() {
+    this.#telnetClient = undefined;
+    try {
+      await this.#_unlock?.();
+    } finally {
+      this.#_unlock = undefined;
+    }
+  }
+
+  /**
+   * Prepares filesystem for writing debug log
+   *
+   * Makes the dirs
+   */
+  async #prepareFs() {
+    await B.all([fs.mkdirp(this.#logDirpath), fs.mkdirp(this.#lockfileDirpath)]);
+  }
+
+  /**
+   * Connects to a debug log and pipes its output somewhere.  Returns a wrapper around the socket.
+   * @returns {Promise<void>} - Telnet client
+   */
+  async connect() {
+    await this.#prepareFs();
+
+    this.#_unlock = await this.#lockLogfile(this.#hostSlug, this.logPath, this.#lockfileDirpath, {
+      fd: this.#opts.logFd,
+    });
+    const client = (this.#telnetClient = new Telnet());
+    await client.connect({
+      host: this.#host,
+      port: this.#opts.port,
+      encoding: 'utf8',
+      ...this.#opts.telnetOpts,
+    });
+    log.debug(`Connected to ${this.#host}`);
+
+    client
+      .on('error', (err) => {
+        log.error(err);
+        try {
+          this.emit('error', err);
+        } finally {
+          this.#cleanup();
+        }
+      })
+      .on('close', () => {
+        log.info(`Connection to host ${this.#host} closed`);
+        try {
+          this.emit('close');
+        } finally {
+          this.#cleanup();
+        }
+      })
+      .on('timeout', () => {
+        // inactivity timeout.  'close' will also be emitted (I think), so don't unlock here
+        this.emit('timeout');
+      });
+
+    this.#pipeToLog();
+  }
+
+  /**
+   * Configure a pipe from the connected telnet client to the log file
+   */
+  #pipeToLog = () => {
+    const sock = this.#telnetClient?.getSocket();
+    if (sock) {
+      // note: the encoding may just be `ascii` since that's what telnet is supposed to be
+      sock
+        .setEncoding('utf8')
+        .on('finish', () => {
+          this.emit('finish');
+        })
+        .on('open', () => {
+          log.info(`Now writing to logfile ${this.logPath}`);
+        })
+        .pipe(_fs.createWriteStream(this.logPath, {fd: this.#opts.logFd, encoding: 'utf8'}));
+
+    } else {
+      throw new ReferenceError('Not connected to host');
+    }
+  };
+
+  /**
+   * Resolves w/ the entire contents of the logfile
+   */
+  async getLog() {
+    return await fs.readFile(this.logPath, 'utf8');
```

Comment:
the way the w3c spec is defined (we should double check), only new logs that have been written since the last request to retrieve logs should be returned. we may be better off using `fs.read` and keeping track of a file pointer?

Reply:
I have an implementation for this, but working on testing to make sure everything places nice together.

I've split it up into separate classes in the name of separation-of-concerns:

- `RokuDebugClient` which is the telnet adapter
- `RokuDebugWriter` which pipes a `RokuDebugClient` into a logfile
- `FileReader` which is a generic class that reads a file piecemeal

If the cap is enabled, the driver will instantiate a `RokuDebugWriter` and a `FileReader`; whenever the "get log" command happens, `FileReader#read()` is called and the contents of the logfile from last read to EOF are returned

Reply:
update: I've written unit tests.  next is E2E tests.  I expect to acquire a device soon which should make things easier. I will probably want to create a sample app too